packages/bits/src/lib/wizard-v2/wizard-horizontal/wizard-horizontal.component.ts

```diff
@@ -41,10 +80,236 @@ export class WizardHorizontalComponent extends WizardDirective implements OnInit
+            this.stepsCachedArray = [...this.stepsArray];
+        });
+
+        this.selectionChange
+            .pipe(takeUntil(this._destroyed))
+            .subscribe((event: StepperSelectionEvent) => {
+
+                // This describes the navigation between the headers in a visible and hidden areas in the responsive mode.
+                if (this.hasOverflow) {
+                    if (event.previouslySelectedIndex < event.selectedIndex) {
+                        if (this.overflownStepsEnd.length) {
+                            const selectedIndex = this.overflownStepsEnd.findIndex((item) => item === event.selectedStep as WizardStepV2Component);
+                            this.visibleSteps = this.visibleSteps.concat(this.overflownStepsEnd.splice(0, selectedIndex + 1));
+                            this.overflownStepsStart = this.overflownStepsStart.concat(this.visibleSteps.splice(0, selectedIndex + 1));
+                        }
+                    } else {
+                        if (this.overflownStepsStart.length) {
+                            const startPiece = this.overflownStepsStart.splice(event.selectedIndex, this.overflownStepsStart.length - event.selectedIndex);
+                            this.visibleSteps = startPiece.concat(this.visibleSteps);
+                            this.overflownStepsEnd.splice(0, 0, ...this.visibleSteps.splice(this.visibleSteps.length - startPiece.length, startPiece.length));
+                        }
+                    }
+                }
+            });
+
+        this.headerResizeObserver.observe(this.headerContainer.nativeElement);
+    }
+
+    public ngOnDestroy(): void {
+        super.ngOnDestroy();
+
+        this.headerResizeObserver.disconnect();
+    }
+
+    private handleOverflow(): void {
+        this.checkOverflow();
+
+        if (this.hasOverflow || this.isInResponsiveMode) {
+            const numberOfVisibleItems = this.numberOfVisibleItems;
+
+            if (!this.isInResponsiveMode) {
+                if (this.state?.finished) {
+                    this.visibleSteps = this.stepsArray.slice(-numberOfVisibleItems);
+                    this.overflownStepsStart = this.stepsArray.slice(0, -numberOfVisibleItems);
+
+                    // This covers an edge case with the saved wizard state. It can happen that the active element
+                    // appears inside the hidden list of items, when the dynamnic steps are re-added.
+                    // We make it visible again by retrieving it back to the visible items array
+                    if (this.overflownStepsStart.length && !this.isCurrentStepVisible) {
+                        do {
+                            this.takeLastAddFirst(this.overflownStepsStart, this.visibleSteps);
+                            this.takeLastAddFirst(this.visibleSteps, this.overflownStepsEnd);
+                        }
+                        while(!this.isCurrentStepVisible)
+                    }
```

Comment:
Can be changed to
`if (this.overflownStepsStart.length) {
          while(!this.isCurrentStepVisible){
                   this.takeLastAddFirst(this.overflownStepsStart, this.visibleSteps);
                   this.takeLastAddFirst(this.visibleSteps, this.overflownStepsEnd);
           }
}`

Reply:
The reason `do - while` fits there better is that if we've entered that if statement, we have to run this code at least once anyway. With the `while` going first we'll need to do the `this.isCurrentStepVisible` check twice in a row which is not ideal.

Reply:
on the other hand, maybe the difference not as big as I thought

Reply:
changed